platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DependencyGraphBuilder.java

```diff
@@ -484,9 +484,7 @@ private void validateMultipleNodeSelection(ModuleResolveState module, ComponentS
             assertCompatibleAttributes(first, second, incompatibleNodes);
         }
         if (!incompatibleNodes.isEmpty()) {
-            IncompatibleArtifactVariantsException variantsSelectionException = new IncompatibleArtifactVariantsException(
-                IncompatibleVariantsSelectionMessageBuilder.buildMessage(selected, incompatibleNodes)
-            );
+            IncompatibleArtifactVariantsException variantsSelectionException = variantSelector.getFailureProcessor().incompatibleArtifactVariantsFailure(selected, incompatibleNodes);
```

Comment:
It seems odd that this accesses the failure processor through the variant selector since this validation step does not depend on variant selection. 

To me, this signals that the failure process encompasses even more than just variant selection. That would mean we should probably rename `SelectionFailureHandler` to something more generic and we should be able to access it without going through the variant selector. 

Alternatively, we can continue to instantiate this exception the way we do currently without going through the selection failure handler.

Reply:
Discussed with Justin, suggestions for renaming and passing just failure handler to validation routines implemented.